### PR TITLE
New DNP3Agent and DNP3 driver modeling a DNP3 outstation

### PR DESCRIPTION
**Pull request by rob-calvert**
# Description

This submission implements a new protocol agent, DNP3Agent, along with a new "dnp3" driver that interacts with DNP3Agent and publishes DNP3 points to the message bus.

The agent and driver depend on an additional Python library, pydnp3. As of now, a nonstandard command sequence must be used to install the library (i.e., not simply a "pip install"). A README.md file in the agent's directory describes how to install the library.

This change resolves Issue #314.

# Type of change

- New feature (non-breaking change which adds functionality).
- This change requires a documentation update. A new ReadTheDocs page describes the agent and driver.

# How Has This Been Tested?

The new agent and driver were tested using automated regression tests (pytests) and a standalone test agent (the pull request includes its code). The pytests are now annotated "mark.skip" because of the dependency on a pydnp3 library that has a nonstandard install — that library wouldn't be available to Travis.
<a href='#crh-start'></a><a href='#crh-data-%7B%22processed%22%3A%20%5B%22https%3A//github.com/VOLTTRON/volttron/pull/1676%23issuecomment-387257986%22%2C%20%22https%3A//github.com/VOLTTRON/volttron/pull/1676%23issuecomment-387262277%22%2C%20%22https%3A//github.com/VOLTTRON/volttron/pull/1676%23issuecomment-387490347%22%2C%20%22https%3A//github.com/VOLTTRON/volttron/pull/1676%23discussion_r186493941%22%2C%20%22https%3A//github.com/VOLTTRON/volttron/pull/1676%23discussion_r186867752%22%2C%20%22https%3A//github.com/VOLTTRON/volttron/pull/1676%23discussion_r186871437%22%2C%20%22https%3A//github.com/VOLTTRON/volttron/pull/1676%23discussion_r186873093%22%2C%20%22https%3A//github.com/VOLTTRON/volttron/pull/1676%23discussion_r187204430%22%2C%20%22https%3A//github.com/VOLTTRON/volttron/pull/1676%23issuecomment-396386917%22%5D%2C%20%22comments%22%3A%20%7B%22General%20Comment%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/VOLTTRON/volttron/pull/1676%23issuecomment-387257986%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22I%20was%20unable%20to%20build%20the%20pydnp3%20library%20on%20either%20of%20my%20vms.%20%20I%20used%20ubuntu%2018.04%20and%20linux%20mint%20and%20both%20of%20them%20ran%20out%20of%20memory%20building%20the%20final%20piece%20of%20code...%5Cr%5Cn%5Cr%5Cn%21%5Bimage%5D%28https%3A//user-images.githubusercontent.com/3979063/39733176-4435cc8e-5226-11e8-90a1-80566bc18935.png%29%5Cr%5Cn%5Cr%5Cn%5Cr%5Cn%21%5Bimage%5D%28https%3A//user-images.githubusercontent.com/3979063/39733126-f66483f6-5225-11e8-97d5-3effe5d8fddd.png%29%5Cr%5Cn%5Cr%5Cn%21%5Bimage%5D%28https%3A//user-images.githubusercontent.com/3979063/39733168-2dfa251e-5226-11e8-97a9-cae86e97e28f.png%29%5Cr%5Cn%22%2C%20%22created_at%22%3A%20%222018-05-08T01%3A41%3A44Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars3.githubusercontent.com/u/3979063%3Fv%3D4%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/craig8%22%7D%7D%2C%20%7B%22body%22%3A%20%22Yes%2C%208Gb%20will%20do%20it.%20%20We%20are%20building%20on%20Ubuntu%2016.04%20with%208Gb.%20%20Final%20library%20size%20is%20only%204.9Mb.%20%20We%20will%20look%20at%20dynamic%20linking%20instead%20of%20static%2C%20which%20might%20reduce%20the%20memory%20required%20for%20build%2C%20but%20not%20sure.%22%2C%20%22created_at%22%3A%20%222018-05-08T02%3A10%3A36Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars3.githubusercontent.com/u/500659%3Fv%3D4%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/ffreckle%22%7D%7D%2C%20%7B%22body%22%3A%20%22Ack%208Gb%20alrighty%20will%20attempt%20that...%22%2C%20%22created_at%22%3A%20%222018-05-08T18%3A03%3A02Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars3.githubusercontent.com/u/3979063%3Fv%3D4%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/craig8%22%7D%7D%2C%20%7B%22body%22%3A%20%22Superseded%20by%20%231695.%22%2C%20%22created_at%22%3A%20%222018-06-11T21%3A04%3A47Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars0.githubusercontent.com/u/11467133%3Fv%3D4%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/rob-calvert%22%7D%7D%5D%2C%20%22title%22%3A%20%22General%20Comment%22%7D%2C%20%22Pull%205d1f115c4337cfb466e72edf03c36970eb55fc0c%20services/core/DNP3Agent/README.md%208%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/VOLTTRON/volttron/pull/1676%23discussion_r186493941%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22Should%20add%20the%20system%20dependencies%20on%20CMake%20and%20possibly%20the%20memory%20requirements%20for%20the%20pydnp3%20wrapper%20etc.%22%2C%20%22created_at%22%3A%20%222018-05-07T17%3A40%3A43Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars3.githubusercontent.com/u/3979063%3Fv%3D4%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/craig8%22%7D%7D%2C%20%7B%22body%22%3A%20%22When%20running%20through%20the%20README.md%20in%20the%20root%20of%20DNP3Agent%20I%20get%20the%20following%20output%20in%20the%20stack%20trace%20in%20the%20volttron%20log.%5Cr%5Cn%5Cr%5Cn%60%60%60%60%20bash%5Cr%5Cn2018-05-08%2013%3A47%3A47%2C038%20%28dnp3agent-1.0%2014628%29%20dnp3.agent%20DEBUG%3A%20DNP3Agent%20configuration%20parameters%3A%5Cr%5Cn2018-05-08%2013%3A47%3A47%2C038%20%28dnp3agent-1.0%2014628%29%20dnp3.agent%20DEBUG%3A%20%5Ctpoint_definitions_path%3D%7E/repos/volttron/services/core/DNP3Agent/opendnp3_data.config%5Cr%5Cn2018-05-08%2013%3A47%3A47%2C038%20%28dnp3agent-1.0%2014628%29%20dnp3.agent%20DEBUG%3A%20%5Ctpoint_topic%3Ddnp3/point%5Cr%5Cn2018-05-08%2013%3A47%3A47%2C038%20%28dnp3agent-1.0%2014628%29%20dnp3.agent%20DEBUG%3A%20%5Ctlocal_ip%3D0.0.0.0%5Cr%5Cn2018-05-08%2013%3A47%3A47%2C038%20%28dnp3agent-1.0%2014628%29%20dnp3.agent%20DEBUG%3A%20%5Ctport%3D20000%5Cr%5Cn2018-05-08%2013%3A47%3A47%2C038%20%28dnp3agent-1.0%2014628%29%20dnp3.agent%20DEBUG%3A%20%5Ctoutstation_config%3D%7B%27log_levels%27%3A%20%5B%27NORMAL%27%2C%20%27ALL_APP_COMMS%27%5D%7D%5Cr%5Cn2018-05-08%2013%3A47%3A47%2C039%20%28dnp3agent-1.0%2014628%29%20dnp3.models%20DEBUG%3A%20Loading%20DNP3%20point%20definitions.%5Cr%5Cn2018-05-08%2013%3A47%3A47%2C039%20%28dnp3agent-1.0%2014628%29%20volttron.platform.vip.agent.subsystems.configstore%20ERROR%3A%20Problem%20processing%20callback%3A%5Cr%5Cn2018-05-08%2013%3A47%3A47%2C040%20%28dnp3agent-1.0%2014628%29%20volttron.platform.vip.agent.subsystems.configstore%20ERROR%3A%20Traceback%20%28most%20recent%20call%20last%29%3A%5Cr%5Cn%20%20File%20%5C%22/home/vm/git/volttron/volttron/platform/vip/agent/subsystems/configstore.py%5C%22%2C%20line%20295%2C%20in%20_process_callbacks_one_config%5Cr%5Cn%20%20%20%20callback%28name_map%5Bconfig_name%5D%2C%20action%2C%20contents%29%5Cr%5Cn%20%20File%20%5C%22/home/vm/.volttron/agents/56b6cd6d-e8f4-42c9-a83d-0f5662329413/dnp3agent-1.0/dnp3/agent.py%5C%22%2C%20line%20140%2C%20in%20_configure%5Cr%5Cn%20%20%20%20self.load_point_definitions%28%29%5Cr%5Cn%20%20File%20%5C%22/home/vm/.volttron/agents/56b6cd6d-e8f4-42c9-a83d-0f5662329413/dnp3agent-1.0/dnp3/agent.py%5C%22%2C%20line%20261%2C%20in%20load_point_definitions%5Cr%5Cn%20%20%20%20PointDefinition.load_points%28expanded_path%29%5Cr%5Cn%20%20File%20%5C%22/home/vm/.volttron/agents/56b6cd6d-e8f4-42c9-a83d-0f5662329413/dnp3agent-1.0/dnp3/models.py%5C%22%2C%20line%20348%2C%20in%20load_points%5Cr%5Cn%20%20%20%20raise%20ValueError%28%27Problem%20parsing%20%7B%7D.%20No%20data%20loaded.%20Error%3D%7B%7D%27.format%28point_definitions_path%2C%20err%29%29%5Cr%5CnValueError%3A%20Problem%20parsing%20/home/vm/repos/volttron/services/core/DNP3Agent/opendnp3_data.config.%20No%20data%20loaded.%20Error%3D%5BErrno%202%5D%20No%20such%20file%20or%20directory%3A%20%27/home/vm/repos/volttron/services/core/DNP3Agent/opendnp3_data.config%27%5Cr%5Cn%5Cr%5Cn%60%60%60%60%22%2C%20%22created_at%22%3A%20%222018-05-08T21%3A06%3A33Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars3.githubusercontent.com/u/3979063%3Fv%3D4%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/craig8%22%7D%7D%2C%20%7B%22body%22%3A%20%22The%20problem%20is%20there%20is%20an%20embed%20repos%20in%20the%20path...mine%20has%20git%20not%20repos%20%3A%29%22%2C%20%22created_at%22%3A%20%222018-05-08T21%3A19%3A59Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars3.githubusercontent.com/u/3979063%3Fv%3D4%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/craig8%22%7D%7D%2C%20%7B%22body%22%3A%20%22Ahh%20IC%20you%20have%20referenced%20the%20currently%20passed%20in%20config%20file%20inside%20the%20config%20file.%20%20Is%20there%20any%20reason%20you%20aren%27t%20using%20the%20config%20store%20for%20this%3F%22%2C%20%22created_at%22%3A%20%222018-05-08T21%3A26%3A32Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars3.githubusercontent.com/u/3979063%3Fv%3D4%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/craig8%22%7D%7D%2C%20%7B%22body%22%3A%20%22Thanks%20for%20the%20comments%2C%20Craig.%20In%20a%20moment%20I%27ll%20commit%20an%20updated%20version%20of%20README.md%20that%20describes%20the%20memory%20requirements%2C%20the%20CMake%20dependency%2C%20and%20the%20need%20to%20adjust%20point_definitions_path%20in%20dnp3agent.config.%20%5Cr%5Cn%5Cr%5CnKisensum%20is%20working%20on%20an%20update%20%28to%20be%20submitted%20as%20a%20separate%20pull%20request%29%20that%20creates%20a%20second%2C%20alternative%20version%20of%20the%20agent%2C%20MesaAgent%2C%20from%20a%20%5C%22mesa%5C%22%20subdirectory%20of%20DNP3Agent/dnp3.%20As%20part%20of%20that%20pull%20request%2C%20we%20plan%20to%20move%20the%20DNP3%20point%20configuration%20file%20into%20the%20VOLTTRON%20config%20store.%22%2C%20%22created_at%22%3A%20%222018-05-09T23%3A30%3A04Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars0.githubusercontent.com/u/11467133%3Fv%3D4%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/rob-calvert%22%7D%7D%5D%2C%20%22title%22%3A%20%22File%3A%20services/core/DNP3Agent/README.md%3AL1-23%22%7D%7D%7D'></a>
<a href='https://www.codereviewhub.com/'></a>
- [ ] <a href='#crh-comment-Pull 5d1f115c4337cfb466e72edf03c36970eb55fc0c services/core/DNP3Agent/README.md 8'></a> &nbsp;<b><a href='https://github.com/VOLTTRON/volttron/pull/1676#discussion_r186493941'>File: services/core/DNP3Agent/README.md:L1-23</a></b>
- <a href='https://github.com/craig8'></a> Should add the system dependencies on CMake and possibly the memory requirements for the pydnp3 wrapper etc.
- <a href='https://github.com/craig8'></a> When running through the README.md in the root of DNP3Agent I get the following output in the stack trace in the volttron log.
```` bash
2018-05-08 13:47:47,038 (dnp3agent-1.0 14628) dnp3.agent DEBUG: DNP3Agent configuration parameters:
2018-05-08 13:47:47,038 (dnp3agent-1.0 14628) dnp3.agent DEBUG: 	point_definitions_path=~/repos/volttron/services/core/DNP3Agent/opendnp3_data.config
2018-05-08 13:47:47,038 (dnp3agent-1.0 14628) dnp3.agent DEBUG: 	point_topic=dnp3/point
2018-05-08 13:47:47,038 (dnp3agent-1.0 14628) dnp3.agent DEBUG: 	local_ip=0.0.0.0
2018-05-08 13:47:47,038 (dnp3agent-1.0 14628) dnp3.agent DEBUG: 	port=20000
2018-05-08 13:47:47,038 (dnp3agent-1.0 14628) dnp3.agent DEBUG: 	outstation_config={'log_levels': ['NORMAL', 'ALL_APP_COMMS']}
2018-05-08 13:47:47,039 (dnp3agent-1.0 14628) dnp3.models DEBUG: Loading DNP3 point definitions.
2018-05-08 13:47:47,039 (dnp3agent-1.0 14628) volttron.platform.vip.agent.subsystems.configstore ERROR: Problem processing callback:
2018-05-08 13:47:47,040 (dnp3agent-1.0 14628) volttron.platform.vip.agent.subsystems.configstore ERROR: Traceback (most recent call last):
File "/home/vm/git/volttron/volttron/platform/vip/agent/subsystems/configstore.py", line 295, in _process_callbacks_one_config
callback(name_map[config_name], action, contents)
File "/home/vm/.volttron/agents/56b6cd6d-e8f4-42c9-a83d-0f5662329413/dnp3agent-1.0/dnp3/agent.py", line 140, in _configure
self.load_point_definitions()
File "/home/vm/.volttron/agents/56b6cd6d-e8f4-42c9-a83d-0f5662329413/dnp3agent-1.0/dnp3/agent.py", line 261, in load_point_definitions
PointDefinition.load_points(expanded_path)
File "/home/vm/.volttron/agents/56b6cd6d-e8f4-42c9-a83d-0f5662329413/dnp3agent-1.0/dnp3/models.py", line 348, in load_points
raise ValueError('Problem parsing {}. No data loaded. Error={}'.format(point_definitions_path, err))
ValueError: Problem parsing /home/vm/repos/volttron/services/core/DNP3Agent/opendnp3_data.config. No data loaded. Error=[Errno 2] No such file or directory: '/home/vm/repos/volttron/services/core/DNP3Agent/opendnp3_data.config'
````
- <a href='https://github.com/craig8'></a> The problem is there is an embed repos in the path...mine has git not repos :)
- <a href='https://github.com/craig8'></a> Ahh IC you have referenced the currently passed in config file inside the config file.  Is there any reason you aren't using the config store for this?
- <a href='https://github.com/rob-calvert'></a> Thanks for the comments, Craig. In a moment I'll commit an updated version of README.md that describes the memory requirements, the CMake dependency, and the need to adjust point_definitions_path in dnp3agent.config.
Kisensum is working on an update (to be submitted as a separate pull request) that creates a second, alternative version of the agent, MesaAgent, from a "mesa" subdirectory of DNP3Agent/dnp3. As part of that pull request, we plan to move the DNP3 point configuration file into the VOLTTRON config store.
- [ ] <a href='#crh-comment-General Comment'></a> &nbsp;<b><a href='https://github.com/VOLTTRON/volttron/pull/1676#issuecomment-387257986'>General Comment</a></b>
- <a href='https://github.com/craig8'></a> I was unable to build the pydnp3 library on either of my vms.  I used ubuntu 18.04 and linux mint and both of them ran out of memory building the final piece of code...
image
image
image
- <a href='https://github.com/ffreckle'></a> Yes, 8Gb will do it.  We are building on Ubuntu 16.04 with 8Gb.  Final library size is only 4.9Mb.  We will look at dynamic linking instead of static, which might reduce the memory required for build, but not sure.
- <a href='https://github.com/craig8'></a> Ack 8Gb alrighty will attempt that...
- <a href='https://github.com/rob-calvert'></a> Superseded by #1695.

<a href='https://www.codereviewhub.com/VOLTTRON/volttron/pull/1676?mark_as_completed=1'></a>&nbsp;<a href='https://www.codereviewhub.com/VOLTTRON/volttron/pull/1676?approve=1'></a>&nbsp;<a href='https://github.com/VOLTTRON/volttron/pull/1676'></a>
<a href='#crh-end'></a>

### CODE REVIEW

**Comment by craig8** [comment]
This pull request is now managed with CodeReviewHub. <a href='https://github.com/VOLTTRON/volttron/pull/1676'>Refresh</a>.<a href='#crh-update-none'></a>

**Comment by craig8** [comment]
I was unable to build the pydnp3 library on either of my vms.  I used ubuntu 18.04 and linux mint and both of them ran out of memory building the final piece of code...

image


image

image

**Comment by ffreckle** [comment]
Yes, 8Gb will do it.  We are building on Ubuntu 16.04 with 8Gb.  Final library size is only 4.9Mb.  We will look at dynamic linking instead of static, which might reduce the memory required for build, but not sure.

**Comment by craig8** [comment]
Ack 8Gb alrighty will attempt that...

File: services/core/DNP3Agent/README.md [hunk 1]
```diff
@@ -0,0 +1,22 @@
+DNP3Agent is a VOLTTRON agent that handles DNP3 outstation communications.
+
+DNP3Agent models a DNP3 outstation, communicating with a DNP3 master.
+
+For further information about this agent and DNP3 communications, please see the VOLTTRON
+DNP3 specification, located in VOLTTRON readthedocs 
+under http://volttron.readthedocs.io/en/develop/specifications/dnp3_agent.html.
+
```
[the remark this answers is not in the document]
Comment (craig8) [reply services/core/DNP3Agent/README.md]: When running through the README.md in the root of DNP3Agent I get the following output in the stack trace in the volttron log.

```` bash
2018-05-08 13:47:47,038 (dnp3agent-1.0 14628) dnp3.agent DEBUG: DNP3Agent configuration parameters:
2018-05-08 13:47:47,038 (dnp3agent-1.0 14628) dnp3.agent DEBUG: 	point_definitions_path=~/repos/volttron/services/core/DNP3Agent/opendnp3_data.config
2018-05-08 13:47:47,038 (dnp3agent-1.0 14628) dnp3.agent DEBUG: 	point_topic=dnp3/point
2018-05-08 13:47:47,038 (dnp3agent-1.0 14628) dnp3.agent DEBUG: 	local_ip=0.0.0.0
2018-05-08 13:47:47,038 (dnp3agent-1.0 14628) dnp3.agent DEBUG: 	port=20000
2018-05-08 13:47:47,038 (dnp3agent-1.0 14628) dnp3.agent DEBUG: 	outstation_config={'log_levels': ['NORMAL', 'ALL_APP_COMMS']}
2018-05-08 13:47:47,039 (dnp3agent-1.0 14628) dnp3.models DEBUG: Loading DNP3 point definitions.
2018-05-08 13:47:47,039 (dnp3agent-1.0 14628) volttron.platform.vip.agent.subsystems.configstore ERROR: Problem processing callback:
2018-05-08 13:47:47,040 (dnp3agent-1.0 14628) volttron.platform.vip.agent.subsystems.configstore ERROR: Traceback (most recent call last):
  File "/home/vm/git/volttron/volttron/platform/vip/agent/subsystems/configstore.py", line 295, in _process_callbacks_one_config
    callback(name_map[config_name], action, contents)
  File "/home/vm/.volttron/agents/56b6cd6d-e8f4-42c9-a83d-0f5662329413/dnp3agent-1.0/dnp3/agent.py", line 140, in _configure
    self.load_point_definitions()
  File "/home/vm/.volttron/agents/56b6cd6d-e8f4-42c9-a83d-0f5662329413/dnp3agent-1.0/dnp3/agent.py", line 261, in load_point_definitions
    PointDefinition.load_points(expanded_path)
  File "/home/vm/.volttron/agents/56b6cd6d-e8f4-42c9-a83d-0f5662329413/dnp3agent-1.0/dnp3/models.py", line 348, in load_points
    raise ValueError('Problem parsing {}. No data loaded. Error={}'.format(point_definitions_path, err))
ValueError: Problem parsing /home/vm/repos/volttron/services/core/DNP3Agent/opendnp3_data.config. No data loaded. Error=[Errno 2] No such file or directory: '/home/vm/repos/volttron/services/core/DNP3Agent/opendnp3_data.config'

````

[the remark this answers is not in the document]
Comment (craig8) [reply services/core/DNP3Agent/README.md]: Ahh IC you have referenced the currently passed in config file inside the config file.  Is there any reason you aren't using the config store for this?

[the remark this answers is not in the document]
Comment (rob-calvert) [reply services/core/DNP3Agent/README.md]: Thanks for the comments, Craig. In a moment I'll commit an updated version of README.md that describes the memory requirements, the CMake dependency, and the need to adjust point_definitions_path in dnp3agent.config. 

Kisensum is working on an update (to be submitted as a separate pull request) that creates a second, alternative version of the agent, MesaAgent, from a "mesa" subdirectory of DNP3Agent/dnp3. As part of that pull request, we plan to move the DNP3 point configuration file into the VOLTTRON config store.

**Comment by rob-calvert** [comment]
Superseded by #1695.

**Pull request closed without merging**